lib/hllcounter.cc

```diff
@@ -0,0 +1,434 @@
+//
+// This file is part of khmer, http://github.com/ged-lab/khmer/, and is
+// Copyright (C) Michigan State University, 2009-2013. It is licensed under
+// the three-clause BSD license; see doc/LICENSE.txt. Contact: ctb@msu.edu
+//
+
+#include "hllcounter.hh"
+
+#include <math.h>
+#include <algorithm>
+#include <numeric>
+#include <inttypes.h>
+#include <sstream>
+
+#include "khmer.hh"
+#include "kmer_hash.hh"
+#include "read_parsers.hh"
+
+#ifdef _OPENMP
+   #include <omp.h>
+#else
+   #define omp_get_thread_num() 0
+   #define omp_get_num_threads() 1
+#endif
+
+#define arr_len(a) (a + sizeof a / sizeof a[0])
+
+using namespace khmer;
+
+std::map<int, std::vector<double> > rawEstimateData;
+std::map<int, std::vector<double> > biasData;
+
+
+double get_alpha(const int p)
+{
+    if ((p < 4) or (p > 16)) {
+        double valid_lower_bound = 1.04 / std::sqrt(std::pow(static_cast<float>(2), 17));
+        double valid_upper_bound = 1.04 / std::sqrt(std::pow(static_cast<float>(2), 3));
+
+        std::stringstream message;
+        if (p < 4) {
+            message << "Max error is " << valid_upper_bound;
+        } else {
+            message << "Min error is " << valid_lower_bound;
+        }
+        throw khmer_exception(message.str().c_str());
+    }
+
+    switch (p) {
+    case 4:
+        return 0.673;
+    case 5:
+        return 0.697;
+    case 6:
+        return 0.709;
+    default:
+        return 0.7213 / (1.0 + 1.079 / (1 << p));
```

Comment:
Nope; you are off the hook.

@ctb do we want to output cite requests for other algos?